src/main/java/org/javarosa/core/model/condition/EvaluationContext.java

```diff
@@ -748,17 +755,56 @@ public EvaluationContext spawnWithCleanLifecycle() {
         return spawnWithCleanLifecycle(null);
     }
 
-    public EvaluationContext spawnWithCleanLifecycle(Map<String, DataInstance> additionalInstances) {
+    public EvaluationContext spawnWithCleanLifecycle(Map<String, ExternalDataInstance> additionalInstances) {
         EvaluationContext ec = new EvaluationContext(this, this.getContextRef());
-        if (additionalInstances != null) {
-            additionalInstances.forEach((name, instance) -> {
-                if (!ec.formInstances.containsKey(name)) {
-                    ec.formInstances.put(name, instance);
-                }
-            });
-        }
         QueryContext qc = ec.getCurrentQueryContext().forceNewChildContext();
         ec.setQueryContext(qc);
+        if (additionalInstances != null) {
+            ec.updateInstances(additionalInstances);
+        }
         return ec;
     }
+
+    private void updateInstances(Map<String, ExternalDataInstance> instances) {
+        Multimap<String, ExternalDataInstance> byRef = getInstancesByRef();
+        instances.forEach((name, instance) -> {
+            String ref = instance.getReference();
+            if (!byRef.containsKey(ref)) {
+                if (formInstances.containsKey(name)) {
+                    throw new RuntimeException(String.format(
+                            "EvaluationContext already contains an instance with ID %s with a different ref", name));
+                }
+                formInstances.put(name, instance);
+            } else {
+                for (ExternalDataInstance existing : byRef.get(ref)) {
+                    if (existing.getRoot() == null) {
+                        // just in time initializing of the instance
+                        String instanceId = existing.getInstanceId();
+                        TreeElement root = (TreeElement)instance.getRoot();
+                        if (instanceId != name) {
+                            root = TreeUtilities.renameInstance(root, instanceId);
+                        }
+                        root.setParent(existing.getBase());
+                        existing.copyFromSource(new ConcreteInstanceRoot(root));
+                    }
+                }
+            }
+            if (!formInstances.containsKey(name) || formInstances.get(name).getRoot() == null) {
+                // instance name is the same so no need to rename it
+                formInstances.put(name, instance);
```

Comment:
should we skip writing to `formInstances` in case there is already an instance in this table with the matching reference to `instance` ? Since one of the conceptual point we are pushing with this redesign is that a reference should uniquely identify the instance, it seems a bit contradictory to me to have multiple instances with same reference but different ids in eval context.

Reply:
This loop only runs after we've already checked against the refs. This is actually a temporary measure to for the existing apps that use the legacy instance names. Once all apps are declaring instances correctly we could remove this entirely.

In the mean time I think this is still safe since we are only replacing instances if they are not already in the context OR they have a null root. I don't think the 2nd condition will actually ever happen so I could look into removing that.

Reply:
that makes sense to me and seems good to go without any further changes. Might make sense to add a comment although to remove the handling around multiple instances once the current apps are all migrated to new source references.